### ISSUE TITLE
cloudabi-run: Launch a graph of connected cloudabi programs?

### ISSUE THREAD

**Issue by laanwj** [NuxiNL/cloudabi-utils#1]
Just a wild idea: I think it could be useful to be able to launch multiple cloudabi programs at the same time, with pipes connecting to each other. Possibly configured from a single YAML file, which also specifies the topology.

This would make it easier to sandbox separate concerns in separate sub-programs. E.g. a process that stores keys, a process that handles incoming network connections, and so on.

Edit: Alternatively I guess this could be handled by having a top-level cloudabi executable that connects and launches the subprograms. However the user would need to give it access to the directory with the binaries. Or pass it pre-opened file-descriptors to the binaries. Yes that would work I guess, with `execveat` with `AT_EMPTY_PATH`.

**Comment by EdSchouten** [comment]
Yes, we should have something like this! YAML supports multiple documents in one file (using `---`). Maybe we could reuse

**Comment by EdSchouten** [comment]
Sorry, accidentally clicked a button before finishing my post.

Yes, having something like this would be really interesting. In the Docker world, there is already something like this called `docker-compose`. I should take a look at that utility to see if I can learn some interesting things from that.

YAML supports multiple documents in one file (using `---`). Maybe this should also form the basis for such a feature?

**Comment by laanwj** [comment]
Docker-compose looks interesting, that's indeed the kind of thing I was thinking about.

> YAML supports multiple documents in one file (using ---). Maybe this should also form the basis for such a feature?

I think so. The challenge here will be, I think, that sockets and other shared resources would have local identifiers that need to be reference-able across documents.

**Comment by EdSchouten** [comment]
Hi Wladimir,

Right now I've got a prototype of running CloudABI software on [Kubernetes](https://kubernetes.io/). With Kubernetes, you can already spawn multiple containers (read: CloudABI programs) at once, by declaring them in the same [pod](https://kubernetes.io/docs/concepts/workloads/pods/pod/). This is also what is implemented by [Scuba](https://github.com/NuxiNL/scuba), our Kubernetes helper daemon.

Long story short, I think it may make sense to leave `cloudabi-run` as is; minimal. Let's keep more advanced ways of starting CloudABI processes to Scuba. Thanks for reporting this issue regardless; it's just that this is being worked on in a way different than anticipated.
